src/base/file_utils.cc

```diff
@@ -188,6 +189,48 @@ ssize_t WriteAll(int fd, const void* buf, size_t count) {
   return static_cast<ssize_t>(written);
 }
 
+base::Status CopyFile(int fd_in, int fd_out) {
+  off_t original_offset = lseek(fd_in, 0, SEEK_CUR);
+  if (original_offset == -1) {
+    return base::ErrStatus(
+        "Can't get offset in 'fd_in', lseek error: %s (errno: %d)",
+        strerror(errno), errno);
+  }
+
+  auto restore_offset_on_exit = OnScopeExit([fd_in, original_offset] {
+    off_t ret = lseek(fd_in, original_offset, SEEK_SET);
+    // 'lseek' should never fail here, but if it fails we crash, to prevent
+    // possible data loss/overwrite in the 'fd_in'.
+    PERFETTO_CHECK(ret >= 0);
+  });
+
+  if (lseek(fd_in, 0, SEEK_SET) == -1) {
+    return base::ErrStatus(
+        "Can't change the offset in 'fd_in', lseek error: %s (errno: %d)",
+        strerror(errno), errno);
+  }
+
+  // Use bigger buffer when copy files.
+  constexpr size_t kCopyFileBufSize = 32 * 1024;  // 32KB.
+  static_assert(kCopyFileBufSize > kBufSize);
+  char buf[kCopyFileBufSize];
```

Comment:
32KB on the stack is too much, especially for a base funciton that could be called eslewhere in future.
Last time I did somethign similar, it started failing on random places (e.g. b/194510398#comment5).
IIRC google3's fibers have a 64K stack by default.

Allocate this buffer on the heap outide of the loop.